### PR TITLE
Refactor index.js to address a few issues 

### LINKED ISSUE #3: elm-package.json is ignored when the elm file to run is in a subdirectory
Hi @jfairbank and thanks for this nice tool! Here's a small issue you might want to have a look at in future.

I've got a few subfolders in my elm projects and I'd like to execute `run-elm /path/to/project/path/to/subfolder/Module.elm`. Despite that there exist `elm-package.json` and `elm-stuff` in `/path/to/project`, these assets are being created in `path/to/subfolder`. When the original `elm-package.json` contains links to non-core third-party modules, `run-elm` fails because a new `elm-package.json` does not include much.

It'd be great if `run-elm` searched for the closest `elm-package.json` before deciding to initialize a new project!

### PR DESCRIPTION
**Pull request by kachkaev**
Closes #3 & #4, possibly addresses #1

### CODE REVIEW

**Comment by kachkaev** [comment]
Ouch, just noticed that you did some refactoring yesterday too @jfairbank! 😭  Anyway, feel free to review my changes to decide what we do next.

My PR should close all open issues. The existing test passes, but I haven't added any new ones yet. Will be happy to do so after your code review.

File: index.js
Comment (kachkaev) [inline index.js]: I know you've already switched to node 8, which supports async/await out of box. However, if we keep this polyfill, we'll be able to support node 6 as well as it's still in maintenance till April 2019.

File: index.js
Comment (kachkaev) [inline index.js]: We can use promisified fs methods from `fs-extra`, so there is no need to use `util` and promisify them manually.

File: index.js
Comment (kachkaev) [inline index.js]: I put all the app's logic into a single async function, which make the data flow much easier to read. With sub-functions (which made sense before switching to async/await) the order of steps becomes a bit less obvious and there's either a need for global variables or for passing misc arguments back and forth.

File: index.js
Comment (kachkaev) [inline index.js]: `commander` npm module (from where we import `program`) gives us an easy way of defining command line args. My PR only supports one argument, but if you want some data to go down to elm like in v2, that should be easy to arrange.

File: index.js
Comment (kachkaev) [inline index.js]: With a timestamp in the end of the main module name, we can avoid naming collisions and race conditions.

File: index.js
Comment (kachkaev) [inline index.js]: This catch block represents all possible failures and thus standardizes error printing.

File: index.js
Comment (kachkaev) [inline index.js]: This option closes #4

File: index.js
Comment (kachkaev) [inline index.js]: This option closes #3

**Review by kachkaev — COMMENTED** [review]
I commented a few bits for you @jfairbank to help you better see the motivation behind my changes. Looking forward to see your review!

File: index.js
Comment (jfairbank) [inline index.js]: Maybe change to:

```
'Runs an Elm file and prints the result of `output` to stdout.'
```

File: index.js
Comment (jfairbank) [inline index.js]:
```
' `output` can be a constant of type `String` or a function that accepts `List String` and returns `String`. You can supply command line arguments to the Elm file if `output` accepts `List String`.'
```

File: index.js
Comment (jfairbank) [inline index.js]:
```
'supply a custom constant/function name different from `output` for printing results`
```

File: index.js
Comment (jfairbank) [inline index.js]:
```
'specific directory to search for elm-package.json or to create it (if different from Elm file location)'
```

File: index.js
Comment (jfairbank) [inline index.js]: Could this validation happen inside the commander config per https://github.com/tj/commander.js/#regular-expression?

Comment (jfairbank) [reply index.js]: Also, we should guard against `main` because the Elm compiler won't accept it if it isn't `Html msg` or `Program ...`

File: index.js
Comment (jfairbank) [inline index.js]: mo**d**ule

File: index.js
Comment (jfairbank) [inline index.js]: It might be worth pulling this outside of the try-catch and throwing an explicit error message that it isn't a file.

File: index.js
Comment (jfairbank) [inline index.js:85]: It might be worth pulling this outside of the try-catch and throwing an explicit error message that it isn't a directory.

File: index.js
Comment (jfairbank) [inline index.js]: I think I wrote the wrong regex here. We need at least one space. Can you change it to this please?

```js
const argsRegex = /^output +\w+ *=/;
```

File: index.js
Comment (jfairbank) [inline index.js]: This should be synchronous already, so we don't need `await`.

File: index.js
Comment (jfairbank) [inline index.js]: We should probably create our own error types to make inspection cleaner here and also below.

**Review by jfairbank — CHANGES_REQUESTED** [review]
Great work @kachkaev!

Thanks for all your effort, especially adding more robust validations and error messages.

I like the overall direction, but I'd really like to go back to the helper function approach from before. Definitely keep all the new code, but can you extract out key parts of the logic into helper functions and let the main async IIFE coordinate them instead?

Also, we'll probably want some tests for the new options. Let me know if you have any trouble with that.

Thanks again!

**Comment by kachkaev** [comment]
Thanks for your review @jfairbank! I've been working on restoring the new feature you've added in 2.0.0; will finish that soon.

Will be also happy to look at splitting the code back into smaller functions. However, given that there're quite a few variable used in various places, I currently doubt that moving several lines of code to different parts of the file will help others read and maintain the code. To be honest I even see this as a step back, because with these extra functions that are called only once anyway, the eye has to jump between various parts of the file instead of sliding smoothly along the flow of chained tasks. These tasks do not have simple i/o (i.e. each of them takes more than one input and sometimes produces more than one output), so switching to helper functions without introducing nasty global vars is quite challenging.

File: index.js
Comment (kachkaev) [reply index.js]: I'll make it even shorter: `constant or function name for printing results`. Help will print:

`--output-name [name]  constant or function name for printing results (default: output)`

File: index.js
Comment (kachkaev) [reply index.js]: I would not rely much on commander. It's a popular module, but it's not working as expected in quite a few scenarios from my experience. E.g. if we specify a regexp for the option and the given string fails the test, the option becomes equal to true and we have to handle that value manually. In any case, our checks are multi-angled and this cannot be fitted into quite basic set of features that commander provides. I'd keep all the checks next to each other and thus protect the reader from sorting out any extra magic that's done outside `index.js` that they have in front of them.

Comment (kachkaev) [reply index.js]: Good point about `main`! We also have a few other things defined in the main elm module, so I'll add `'init', 'main', 'program', 'sendOutput'` to the list of restricted output names.

File: index.js
Comment (kachkaev) [reply index.js]: I'm not sure what for. Current implementation will print you `Error: File '/invalid-path' does not exist`, which looks totally fine to me. I don't see much logic difference between not finding a file,  not finding a project folder, having problems with `output` name or failing to compile the module. They all indicate a problem with the user input and thus can be treated equally. One try / catch for all hiccups that a user may be responsible for.

File: index.js
Comment (kachkaev) [reply index.js:85]: Currently printing `Error: Provided --project-dir '/invalid-dir' is not a directory`. See my response to your prev comment.

File: index.js
Comment (kachkaev) [reply index.js]: Sure! I did not notice that typo too, thanks!

File: index.js
Comment (kachkaev) [reply index.js]: 👌 you're right!

File: index.js
Comment (kachkaev) [reply index.js]: Maybe, but I'm not sure if this is important at the moment. Both to the "special" errors that we handle in a peculiar way are easy to understand ATM. If we add our own error types, the code may look a bit more organised, but will also introduce new abstractions that a reader would have to understand. The simpler the better IMO, we don't want to overengineer.

**Comment by kachkaev** [comment]
@jfairbank I updated a few more things that include responses to your review. Also see my comments to your review comments 😃 Thanks for your quick response again! 🎉 

Feel free to check the docs I've added and review everything again. Once we agree about the new implementation, I'm happy to add a few test cases to cover the new features. Guess the new release can be `2.1.0` since we don't break anything.

What benefits do you see in returning to the helper functions? I'd like to understand what perks may outplay the added complexity mentioned in the previous comment.

**Comment by kachkaev** [comment]
I'm thinking of adding more test later on this week. What do you think of switching to some testing framework such as [jest](https://facebook.github.io/jest/)? That could make the code a bit more organised and also make it easier to add more test cases in future. Just want to hear your opinion before writing anything.

Here are minor issue I see in `run-tests` now:
* hard to add tests for _expected_ failures
* in order to add tests for `--help` and `--version` options, some restructuring might be required
* execution is synchronous → slower performance when there are more folders with tests
* no easy way to calculate code coverage
* may be hard to read for new maintainers, especially after we add more test cases

File: index.js
Comment (jfairbank) [reply index.js]: I actually forgot that these errors are coming from the Elm compiler anyway, so no worries on this one then.

File: index.js
Comment (jfairbank) [inline index.js]: `Array.prototype.includes` should be available in Node 6.

**Comment by jfairbank** [comment]
> What benefits do you see in returning to the helper functions? I'd like to understand what perks may outplay the added complexity I mentioned above.

Organization is a large part of it. Each helper function can encapsulate an important step in the flow and the main function can call them. If those functions are well named, then you can still clearly see the steps taking place with less noise by not having the code all in one place. I had planned on eventually refactoring the code into helper modules/functions to wrangle all the "global" variables, but I haven't had time for that. We can leave this as is for now, but this is something I eventually want to tackle.

> I'm thinking of adding more test later on this week. What do you think of switching to some testing framework such as jest? That could make the code a bit more organised and make it easier to add more test cases in future. Just want to hear your opinion before writing anything.

This would be great to use something more robust. The current testing framework was a minimal starting point. A lot of the pain points you bring up could be more easily tested if most of the business logic were in pure functions too, hence why I eventually want a well-factored implementation with helper modules/functions. I'd say let's minimally get the existing tests into Jest and then see about testing "failures". Let's do a separate PR for it too.

Thanks again for you hard work!

**Review by jfairbank — CHANGES_REQUESTED** [review]

File: CHANGELOG.md
Comment (jfairbank) [inline CHANGELOG.md]: Add `--project-dir=/path/to/project` to help Elm pick the correct `elm-package.json` and `elm-stuff` if the Elm file is in a subdirectory.

File: CHANGELOG.md
Comment (jfairbank) [inline CHANGELOG.md]: Specify a custom name for the output constant or function by adding `--output-name=myCustomConstantOrFunction`.

File: CHANGELOG.md
Comment (jfairbank) [inline CHANGELOG.md]: Add period

File: README.md
Comment (jfairbank) [inline README.md]: Customizing Output Name

File: README.md
Comment (jfairbank) [inline README.md]: Specify a custom name for the output constant or function by adding `--output-name=myCustomConstantOrFunction` to the command:

File: README.md
Comment (jfairbank) [inline README.md]: Running Modules in Elm Project Subdirectories

File: README.md
Comment (jfairbank) [inline README.md]: `elm-run` -> `run-elm`

**Review by jfairbank — CHANGES_REQUESTED** [review]

**Comment by kachkaev** [comment]
> I had planned on eventually refactoring the code into helper modules/functions to wrangle all the "global" variables

Right. If it comes to separate isolated modules with functions, then I agree that extracting these out of the IIFE would be a good thing. Such functions don't rely on any global variables, can be unit-tested or even turned into their own npm packages. I also agree with you that we can keep things as is for now with an intent to start this kind of refactoring in future releases. Meanwhile, the IIFE with all the ’meat‘ in it is ≈150 LOC, which is manageable, although not ideal.

Good that you agree about switching to a testing framework — I'll be happy to help with that. A quick thought: what if in this PR we simply leave testing more or less as is and just add two new folders to cover new options? Once 2.1.0 is released, we'll have time to approach testing properly while others can start enjoying the new stuff.

And finally thanks for your new comments! Will work on them now 👍

File: README.md
Comment (kachkaev) [reply README.md]: 👌 although my British spell checker does not like this 😃

File: README.md
Comment (kachkaev) [reply README.md]: OK. ‘Titleizing’ other headers too.

File: index.js
Comment (kachkaev) [reply index.js]: True: http://node.green/#ES2016-features-Array-prototype-includes

**Comment by kachkaev** [comment]
Done. Will have a quick look at adding two test cases and guess we're done for now.

At what point do you want to replace _next_ in the changelog with 2.1.0 as well as updating the version in package.json? Is this a part of my PR or a commit by you on master just before the release?

File: .gitignore
Comment (kachkaev) [inline .gitignore:8]: We need to keep `elm-packge.json` for this test because it has a custom value for `"source-directories"`

File: index.js
Comment (kachkaev) [inline index.js:120]: A regexp is now used here because `{output}` needs to be replaced twice in both templates

File: index.js
Comment (kachkaev) [inline index.js:101]: I completely forgot to replace `/^${outputName} +\\w+ *=/` with a more general regexp and only discovered this after the new test failed 😆

File: tests/args/input.txt
Comment (kachkaev) [inline tests/args/input.txt:1]: Because `run-elm` input is a bit more sophisticated now, I switched to a compulsory `input.txt` in each test folder instead of a former optional `args.json`

File: tests/run-tests.js
Comment (kachkaev) [inline tests/run-tests.js:29]: `cwd` is being set to allow for relative paths in command arguments (e.g. `Main.elm` instead of `/path/to/Main.elm`)

File: tests/run-tests.js
Comment (kachkaev) [inline tests/run-tests.js:39]: I changed the type of these two variables from Buffer to String to call `toString()` a bit less often

File: tests/subdirectory/elm-package.json
Comment (kachkaev) [inline tests/subdirectory/elm-package.json:7]: FYI `tests/subdirectory/elm-package.json` has been added to git because of this row

**Review by kachkaev — COMMENTED** [review]
🏁

**Comment by kachkaev** [comment]
@jfairbank what else would you like me to look at before this PR can be merged?

**Review by jfairbank — APPROVED** [review]

**Comment by jfairbank** [comment]
Looks good @kachkaev. Thanks again for putting this PR together.

I'll take care of updating the release notes in a separate commit.

**Comment by kachkaev** [comment]
Awesome! Really looking forward to seeing a new release — just want to share the npm package with colleagues!

I can also find time to look into jest this week after the new version is out.

**Comment by jfairbank** [comment]
> I can also find time to look into jest this week after the new version is out.

Sounds great. New release is now out.

**Pull request merged**
